### ISSUE TITLE
twisted.internet.abstract.FileDescriptor says implements(IProducer) but means implements(IPushProducer)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4386]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4386|
|Type|defect|
|Created|2010-03-29 22:43:08Z|
|Branch|https://github.com/twisted/twisted/tree/filedescriptor-interfaces-4386|

This leads to some very unfortunate gaffes, like this one:

```
>>> from twisted.internet.tcp import Server
>>> from twisted.internet.interfaces import IPushProducer
>>> IPushProducer.implementedBy(Server)
False
```

`tcp.Server` is very definitely supposed to implement `IPushProducer`; if it doesn't, it would be nearly impossible to implement any _other_ push producers.

Attachments:

* [demo.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5a/e5ab5daa878977866193850eb40c14b6eda10080/276c9d5ed588cb65de36d4b7dd8ee8220e3791cd.patch) (1165 bytes) - added by moijes12 on 2012-02-15 05:10:22Z - This is a demp patch.
* [4386.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5a/e5ab5daa878977866193850eb40c14b6eda10080/4c21055650c414038311771f23336f1fa67d464b.patch) (1969 bytes) - added by moijes12 on 2012-02-16 07:43:17Z - Please find attached the patch. I felt I should have created class FiledescriptorTests which contains  tests for writeSequence and checks for implementing IPushProducer. Would something like that be ok ?
* [4386.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5a/e5ab5daa878977866193850eb40c14b6eda10080/ef0e3c3b83dc54652478441a0fb2efb2f28559fb.patch) (2078 bytes) - added by moijes12 on 2012-02-16 12:21:05Z - Here's another one.
* [4386.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5a/e5ab5daa878977866193850eb40c14b6eda10080/22a466b1da7b1e5bd0b2a101293428de4678ac0e.patch) (4037 bytes) - added by moijes12 on 2012-02-18 16:43:40Z - Please find attached the patch. However, the tests for the changes made to iocpreactor were skipped so I'm not sure how they will behave. If anyone can, please tell me what I require to run the tests. 
* [4386.patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5a/e5ab5daa878977866193850eb40c14b6eda10080/20ebecc50db6613c972fad4fcb7839095c9b597e.patch) (892 bytes) - added by moijes12 on 2012-03-15 19:05:21Z - All calls to verifyClass only changed assertTrue(verifyClass(...))

<details><summary>Searchable metadata</summary>

```
trac-id__4386 4386
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_filedescriptor_interfaces_4386 branches/filedescriptor-interfaces-4386
branch_author__habnabit habnabit
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1269902588000000 1269902588000000
changetime__1331842721000000 1331842721000000
version__None None
owner__wsanchez wsanchez

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_3"></a>|@moijes12 commented|
|-|-|

How about something like the patch. 

One query :
I've only added a test to check if Server implements IPushProducer. Should tests be added for every class that directly or indirectly inherites abstract.FileDescriptor ?

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_4"></a>|@moijes12 commented|
|-|-|

I meant "demo" and not "demp" patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @moijes12|
|-|-|

Thanks for working on this.  Please add the review keyword when you're ready for feedback, even if you're not certain the patch is perfect.

  1. I think we'll be okay with just a unit test for `FileDescriptor` implementing this interface.  It wouldn't be terrible to make a similar assertion for every other class which is also supposed to implement `IPushProducer`, but we can rely on inheritance to work properly, so if `FileDescriptor` implements the proper interface so will all of its subclasses.
  1. The test should use `zope.interface.verify.verifyClass` or `zope.interface.verify.verifyObject` instead.
  1. Since it will be a test for `FileDescriptor`, it can go in `twisted/internet/test/test_filedescriptor.py`.
  1. This merits a [bug fix news fragment](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Thanks.

> I felt I should have created class FiledescriptorTests which contains tests for writeSequence and checks for implementing IPushProducer.

I'm satisfied with this change, yes.  If `FileDescriptorWriteSequenceTests` had contained more tests, or had contained only tests for `FileDescript.writeSequence`, then I would have recommended leaving it alone and adding a brand new `TestCase` subclass for the new test method.  However, since it only had one `writeSequence` test, and it also had a `write` test, I think it was misnamed, and it works better as a general container for `FileDescriptor` tests.

  1. `test_fileDescriptorImplementsIPushProducer` is a little verbose, though.  It's already clear we're testing `FileDescriptor`, since the test method belongs to `FileDescriptorTests`.
  1. The news fragment should be a `.bugfix` fragment instead of a `.misc` fragment.  This is a nice fix and we'll want it to show up in the news file.  `.misc` fragments just have ticket numbers listed, they don't get any descriptive text.
  1. `twisted.internet.iocpreactor.abstract.FileHandle` has the same bug.  It implements `IPushProducer` but only claims to implement `IProducer`.  Can you add a similar fix for it?

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_10"></a>|@habnabit commented|
|-|-|

(In [33691]) Branching to 'filedescriptor-interfaces-4386'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_11"></a>|@habnabit commented|
|-|-|

(In [33692]) Applying patch from ticket.

Refs [#4386](https://github.com/twisted/twisted/issues/4386).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_12"></a>|@habnabit set owner to @moijes12|
|-|-|

Tests pass on iocpreactor, but the tests in `t.i.test.test_iocp` should be doing `self.assertTrue(verifyClass(...))` instead of just `verifyClass(...)`.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_13"></a>|@moijes12 removed owner|
|-|-|

Made the changes to twisted.internet.test_iocp.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

4386.3 plus the incremental patch looks good.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez set status to `closed`|
|-|-|

(In [33856]) Apple 4386.3.patch and 4386.patch.patch from [[#4386](https://github.com/twisted/twisted/issues/4386)](https://github.com/twisted/twisted/issues/4386): FileDescriptor says implements(IProducer) but means implements(IPushProducer)

Author: moijes12
Reviewer: wsanchez
Fixes: #4386
